aloha-core/src/main/scala/com/eharmony/aloha/dataset/vw/multilabel/VwDownsampledMultilabelRowCreator.scala

```diff
@@ -0,0 +1,378 @@
+      s"nanotime: $time, ip: $ip, threadId: $threadId, threadName: $threadName"
+    )
+
+    scrambled
+  }
+
+  /**
+    * Given an `a` and some `seed`, produce output, including a new seed.
+    *
+    * When using this function, the user is responsible for keeping track of,
+    * and providing the seeds.
+    *
+    * The implementation of this function should be referentially transparent.
+    *
+    * @param a     input
+    * @param seed the random seed which is updated on each call.
+    * @return a tuple where the first element is a Tuple2 whose first element is
+    *         missing and error information and second element is an optional result.
+    *         The second element of the outer Tuple2 is the new state.
+    */
+  override def apply(a: A, seed: Long): ((MissingAndErroneousFeatureInfo, Option[Array[String]]), Long) = {
+    val (missingAndErrs, features) = featuresFunction(a)
+
+    // Get the lazy val once.
+    val labToInd = labelToInd
+
+    // TODO: This seems like it could be optimized.
+    // The positiveLabelsFunction is invoked once and all labels are produced.
+    // Then a set is produced that is used to partition all of the labels in
+    // the training data into positive and negative.  It seems like a slightly
+    // more efficient way to do this would be to create a sorted array of positives
+    // indices.  Then a negative index array could be constructed with the appropriate
+    // size based on the size of allLabelsInTrainingSet and the size of the positives
+    // array.  Next allLabelsInTrainingSet.indices and the positiveIndices array could
+    // be iterated over simultaneously and if the current index in
+    // allLabelsInTrainingSet.indices isn't in the positiveIndices array, then it must
+    // be in the negativeIndices array.  Then then offsets into the two arrays are
+    // incremented.  Both the current and proposed algorithms are O(N), where
+    // N = allLabelsInTrainingSet.indices.size.  But the proposed algorithm would likely
+    // have much better constant factors.
+    //
+    // Note that labels produced by positiveLabelsFunction that are not in the
+    // allLabelsInTrainingSet are discarded without notice.
+    //
+    // TODO: Should this be sci.BitSet?
+    val positiveIndices: Set[Int] = positiveLabelsFunction(a).flatMap(labToInd.get)(breakOut)
+
+    val (vwInput, newSeed) = sampledTrainingInput(
+      features,
+      allLabelsInTrainingSet.indices,
+      positiveIndices,
+      defaultNamespace,
+      namespaces,
+      classNs,
+      dummyClassNs,
+      negativeDummyStr,
+      positiveDummyStr,
+      seed,
+      numDownsampledNegLabels
+    )
+
+    ((missingAndErrs, Option(vwInput)), newSeed)
+  }
+}
+
+
+object VwDownsampledMultilabelRowCreator extends Rand {
+
+  // Expose initSeedScramble to companion class.
+  private def scramble(initSeed: Long): Long = initSeedScramble(initSeed)
```

Comment:
Why bother to have a function that is just a pass through?

Reply:
It needs it because of the way that Scala does mix-in composition.  If you removed it and tried to call `initSeedScramble` directly, the calling code wouldn't be able to access it.  `Rand` could have been mixed into both the class and companion object but I didn't like that either.